Allow users to register their own disabling components / default query filters

alice-i-cecile opened the pull request:
# Objective

Currently, default query filters, as added in #13120 / #17514 are hardcoded to only use a single query filter.

This is limiting, as multiple distinct disabling components can serve important distinct roles. I ran into this limitation when experimenting with a workflow for prefabs, which don't represent the same state as "an entity which is temporarily nonfunctional".

## Solution

1. Change `DefaultQueryFilters` to store a SmallVec of ComponentId, rather than an Option.
2. Expose methods on `DefaultQueryFilters`, `World` and `App` to actually configure this.
3. While we're here, improve the docs, write some tests, make use of FromWorld and make some method names more descriptive.

## Follow-up

I'm not convinced that supporting sparse set disabling components is useful, given the hit to iteration performance and runtime checks incurred. That's disjoint from this PR though, so I'm not doing it here. The existing warnings are fine for now.

## Testing

I've added both a doc test and an mid-level unit test to verify that this works!

cBournhonesque commented:
I also need this, but wasn't there concerns in the previous PR that forces us to abandon letting users define their own disabled components?

alice-i-cecile commented:
> Does this automatically work with scenes?

Can you say more about what you mean?

alice-i-cecile commented:
On sparse set filters: https://github.com/bevyengine/bevy/pull/13120#issuecomment-2216835468 and https://github.com/bevyengine/bevy/pull/13120#issuecomment-2216914231. However, the soundness issues appear to be resolved: https://github.com/bevyengine/bevy/pull/13120#issuecomment-2311426988

That said, I'm happy to just forbid them: I think they're a serious performance footgun.

From @NiseVoid on [Discord](https://discord.com/channels/691052431525675048/749335865876021248/1338556248412651562):

> This brings back the reason that blocked DQF at the end, it encourages people to create multiple components for Disabled and you won't be able to see them again unless you disable all filters, which has other annoying side effects (picking up prefabs as well for example)
> The description also contradicts the implementation, it says "prefab" doesn't "disable", yet as far as the code is concerned it totally does disable
> The naming of methods and the use of a HashSet rather than a Vec further suggest that adding mountains of disabling components is the desired use of the feature 🤔

A Vec is likely faster / more sensible, or even a SmallVec. I don't expect more than a handful of these to exist.

I'll also clean up the description.

We can discuss the ecosystem concerns further: I share some of them, but I also think they're broadly hypothetical. Crates can already effectively have "private entities" by making their component types public or nesting them in a subworld or something: that hasn't been a serious problem AFAIK.

NiseVoid commented:
I think the problem is largely a documentation and method naming issue fwiw.

There are cases where the concerns would apply, like if someone creates dozens of alternatives to `Disabled`, users and other crate will struggle to find that entity when they want to. But cases like `Prefab` or something like an `IsResource` filter would have no overlap with `Disabled`, and thus don't create this problem.

MrGVSV commented:
> > Does this automatically work with scenes?
> 
> 
> 
> Can you say more about what you mean? Scenes aren't hooked up to any prefab system yet for example.

Iirc the original PR to add `Disabled` auto filtered it out of saving to scenes. For custom disabling components, users may also want that behavior (or a way to opt out of it).

alice-i-cecile commented:
> Iirc the original PR to add Disabled auto filtered it out of saving to scenes. For custom disabling components, users may also want that behavior (or a way to opt out of it).

Looking at #17514, this PR actually opts `DefaultQueryFilters` out of being saved to scenes as a resource. Which is sensible enough but doesn't pose any new problems for this API :)  See https://github.com/bevyengine/bevy/pull/17514/files#diff-978e33bfbd643635947d2a46e239e7cc222d75541cd32fd8de6cd4f7fe1ab19dR354

alice-i-cecile commented:
Alright, I've added more documentation, cleaned up the PR description, swapped to a SmallVec and made a nice little Default -> FromWorld change. I think this is in a good spot to merge, but y'all are the reviewers.

@SkiFire13, could I get your review here as well? You had concerns about the original DQF, and I want to be sure that you're okay with the tradeoffs we're making here.

alice-i-cecile commented on crates/bevy_ecs/src/entity_disabling.rs:
```diff
@@ -24,54 +70,132 @@
+/// Think carefully about whether you need to use a new disabling component,
+/// and clearly communicate their presence in any libraries you publish.
+#[derive(Resource, Debug)]
 #[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
```
This is failing in CI now that I'm using a SmallVec. I'd appreciate tips on the best way to fix it.

Carter0 commented on crates/bevy_ecs/src/entity_disabling.rs:
```diff
@@ -24,54 +70,132 @@
+/// Semantically, this component is used to mark entities that are temporarily disabled (typically for gameplay reasons),
+/// but will likely be re-enabled at some point.
+///
+/// Like all disabling components, this only disables the entity itself,
```
For all lot of these comments here, I think you can just reference what you wrote at the top of the module. 

Might be good to build the expectation for users that they should look to the top of the module for most of the comments. That's where most of the info is likely to be I suspect, so that way users can get all the info rather than bits and pieces. Also it prevents us from having to maintain all these comments if they go out of date.

alice-i-cecile replied:
I think that's a good strategy in general! In this particular case, there are enough concerns that I want to go overkill on the warnings. I'd really like to avoid having to take this toy away in the future 😅

SkiFire13 reviewed (commented):
I'm a bit out of loop and recently got a new job so I don't have lot of time to fully review the code. I gave a quick glance and left a couple of comments, from what I can see nothing much has changed in terms of tradeoffs, so I think I have the same concerns as before, but if you're ok with them you could give this a try.

urben1680 commented:
I would like this feature, crate owners can disable entities "for their own reasons" that cannot be trivially enabled by other parts of the code.

alice-i-cecile commented on crates/bevy_ecs/src/entity_disabling.rs:
````diff
@@ -1,21 +1,67 @@
+//!
+//! #[derive(Component)]
+//! struct A;
+//!
+//! let mut world = World::new();
+//! world.register_disabling_component::<Prefab>();
+//! world.spawn((A, Prefab));
+//! world.spawn((A,));
+//! world.spawn((A,));
+//!
+//! let mut normal_query = world.query::<&A>();
+//! assert_eq!(2, normal_query.iter(&world).count());
+//!
+//! let mut prefab_query = world.query_filtered::<&A, With<Prefab>>();
+//! assert_eq!(1, prefab_query.iter(&world).count());
+//!
+//! let mut maybe_prefab_query = world.query::<(&A, Has<Prefab>)>();
+//! assert_eq!(3, maybe_prefab_query.iter(&world).count());
+//! ```
+//!
+//! ## Default query filters
+//!
+//! In Bevy, entity disabling is implemented through the construction of a global "default query filter".
+//! Queries which do not explicitly mention the disabled component will not include entities with that component.
+//! If an entity has multiple disabling components, it will only be included in queries that mention all of them.
+//!
+//! For example, `Query<&Position>` will not include entities with the [`Disabled`] component,
+//! even if they have a `Position` component,
+//! but `Query<&Position, With<Disabled>>` or `Query<(&Position, Has<Disabled>)>` will see them.
+//!
+//! Entities with disabling components are still present in the [`World`] and can be accessed directly,
+//! using methods on [`World`] or [`Commands`](crate::prelude::Commands).
 //!
-//! ### Note
+//! ### Warnings
 //!
-//! Currently only queries for which the cache is built after enabling a filter will have entities
+//! Currently, only queries for which the cache is built after enabling a default query filter will have entities
 //! with those components filtered. As a result, they should generally only be modified before the
 //! app starts.
````
We can add a check for this and an internal bool. Lemme toss that behind a debug asserts flag...

alice-i-cecile commented on crates/bevy_ecs/src/entity_disabling.rs:
```diff
@@ -1,21 +1,67 @@
+//! Queries which do not explicitly mention the disabled component will not include entities with that component.
+//! If an entity has multiple disabling components, it will only be included in queries that mention all of them.
+//!
+//! For example, `Query<&Position>` will not include entities with the [`Disabled`] component,
+//! even if they have a `Position` component,
+//! but `Query<&Position, With<Disabled>>` or `Query<(&Position, Has<Disabled>)>` will see them.
```
Has matches both!

SkiFire13 replied:
Ah right. It still feels kinda hacky though

chescock replied:
Yeah, it's awkward to have an extra `bool` that isn't used, and it will count the component as archetypal access, which is somewhat misleading.  

Would it make sense to have a `QueryFilter` with the meaning "include entities whether or not they have `T`, even though it's normally excluded by default query filters"?  (That wouldn't be part of this PR, of course.)

cBournhonesque reviewed (approved):
Great docs!

alice-i-cecile commented on crates/bevy_ecs/src/entity_disabling.rs:
```diff
@@ -24,54 +70,132 @@
 pub struct DefaultQueryFilters {
-    disabled: Option<ComponentId>,
+    // We only expect a few components per application to act as disabling components, so we use a SmallVec here
+    // to avoid heap allocation in most cases.
```
Won't iterating over the vector be slower than if we use a smallvec? I'm less concerned about the initial allocation than I am about maximizing iteration speed at tiny n.
